### PR TITLE
Call the right cleanup function

### PR DESCRIPTION
**Pull request by ash2k**
**What this PR does / why we need it**:
`defer cleanup()` will always call the function that was returned by the first call to `r.resyncChan()` but it should call the one returned by the last call.

**Special notes for your reviewer**:
This will print `c1`, not `c2`. See https://play.golang.org/p/FDjDbUxOvI
```go
func main() {
	var c func()
	c = c1
	defer c()
	c = c2
}

func c1 () {
	fmt.Println("c1")
}

func c2 () {
	fmt.Println("c2")
}
```

**Release note**:
```release-note
NONE
```
/kind bug
/sig api-machinery

### CODE REVIEW

**Comment by ash2k** [comment]
/assign @sttts 
PTAL, easy review.

File: staging/src/k8s.io/client-go/tools/cache/reflector.go [hunk 1]
```diff
@@ -272,6 +270,10 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 	cancelCh := make(chan struct{})
 	defer close(cancelCh)
 	go func() {
+		resyncCh, cleanup := r.resyncChan()
+		defer func() {
+			cleanup() // Call the current cleanup() function
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/reflector.go]: better: the last one written into `cleanup`

**Comment by sttts** [comment]
a nit about the comment. Feel free to self-lgtm after fixing.

**Comment by ash2k** [comment]
@sttts nit addressed. I cannot approve/lgtm - don't have any permissions on this repo.

**Pull request merged**